src/Middleware/HttpLogging/src/W3CLoggingMiddleware.cs

```diff
@@ -0,0 +1,261 @@
+        internal static readonly int _userNameIndex = BitOperations.Log2((int)W3CLoggingFields.UserName);
+        internal static readonly int _serverNameIndex = BitOperations.Log2((int)W3CLoggingFields.ServerName);
+        internal static readonly int _serverIpIndex = BitOperations.Log2((int)W3CLoggingFields.ServerIpAddress);
+        internal static readonly int _serverPortIndex = BitOperations.Log2((int)W3CLoggingFields.ServerPort);
+        internal static readonly int _methodIndex = BitOperations.Log2((int)W3CLoggingFields.Method);
+        internal static readonly int _uriStemIndex = BitOperations.Log2((int)W3CLoggingFields.UriStem);
+        internal static readonly int _uriQueryIndex = BitOperations.Log2((int)W3CLoggingFields.UriQuery);
+        internal static readonly int _protocolStatusIndex = BitOperations.Log2((int)W3CLoggingFields.ProtocolStatus);
+        internal static readonly int _timeTakenIndex = BitOperations.Log2((int)W3CLoggingFields.TimeTaken);
+        internal static readonly int _protocolVersionIndex = BitOperations.Log2((int)W3CLoggingFields.ProtocolVersion);
+        internal static readonly int _hostIndex = BitOperations.Log2((int)W3CLoggingFields.Host);
+        internal static readonly int _userAgentIndex = BitOperations.Log2((int)W3CLoggingFields.UserAgent);
+        internal static readonly int _cookieIndex = BitOperations.Log2((int)W3CLoggingFields.Cookie);
+        internal static readonly int _refererIndex = BitOperations.Log2((int)W3CLoggingFields.Referer);
+
+        // Number of fields in W3CLoggingFields - equal to the number of _*Index variables above
+        internal const int _fieldsLength = 17;
+
+        /// <summary>
+        /// Initializes <see cref="W3CLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="options"></param>
+        /// <param name="w3cLogger"></param>
+        public W3CLoggingMiddleware(RequestDelegate next, IOptionsMonitor<W3CLoggerOptions> options, W3CLogger w3cLogger)
+        {
+            if (next == null)
+            {
+                throw new ArgumentNullException(nameof(next));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (w3cLogger == null)
+            {
+                throw new ArgumentNullException(nameof(w3cLogger));
+            }
+
+            _next = next;
+            _options = options;
+            _w3cLogger = w3cLogger;
+        }
+
+        /// <summary>
+        /// Invokes the <see cref="HttpLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>HttpResponseLog.cs
+        public async Task Invoke(HttpContext context)
+        {
+            var options = _options.CurrentValue;
+
+            var elements = new string[_fieldsLength];
```

Comment:
nit: stackalloc

Reply:
I don't think you can stackalloc arrays of reference types.

Reply:
>I don't think you can stackalloc arrays of reference types.

Doh, Array pool then 😄 

>Also, this array is still referenced after the method exits.

It isn't, gets converted into a string then pushed to the logger. Although looks like there was some discussion about changing a bunch of the formatting to do it later, so maybe this should wait.